File: app/src/test/java/com/test/xyz/demo/ui/repodetails/mvp/RepoDetailsPresenterTest.java
package com.test.xyz.demo.ui.repodetails.mvp;

import com.test.xyz.demo.domain.interactor.MainInteractor;
import com.test.xyz.demo.domain.repository.api.model.Repo;
import com.test.xyz.demo.ui.BasePresenterTest;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

public class RepoDetailsPresenterTest  extends BasePresenterTest {
    private static final String USER_NAME = "google";
    private static final String PROJECT_ID = "test";

    private RepoDetailsPresenter repoDetailsPresenter;

    @Mock
    private MainInteractor mainInteractor;

    @Mock
    private RepoDetailsView repoDetailsView;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        mockInteractor(mainInteractor);
        repoDetailsPresenter = new RepoDetailsPresenterImpl(repoDetailsView, mainInteractor);
    }

    @Test
    public void requestRepoDetails_shouldReturnRepoDetails() {
        //GIVEN
        //NOTHING

        //WHEN
        repoDetailsPresenter.requestRepoDetails(USER_NAME, PROJECT_ID);

        //THEN
        verify(repoDetailsView).showRepoDetails(any(Repo.class));
        verify(repoDetailsView, never()).showError(any(String.class));
    }

    @Test
    public void requestRepoDetails_whenUserNameIsEmpty_shouldReturnError() {
        //GIVEN
        //NOTHING

        //WHEN
        repoDetailsPresenter.requestRepoDetails("", PROJECT_ID);

        //THEN
        verify(repoDetailsView, never()).showRepoDetails(any(Repo.class));
        verify(repoDetailsView).showError(nullable(String.class));
    }
}

File: app/src/test/java/com/test/xyz/demo/ui/repodetails/vp/RepoDetailsPresenterTest.kt
@file:Suppress("IllegalIdentifier")

package com.test.xyz.demo.ui.repodetails.vp

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.never
import com.nhaarman.mockito_kotlin.verify
import com.test.xyz.demo.domain.interactor.MainInteractor
import com.test.xyz.demo.domain.repository.api.model.Repo
import com.test.xyz.demo.ui.BasePresenterTest
import org.junit.Before
import org.junit.Test

class RepoDetailsPresenterTest : BasePresenterTest() {

    lateinit var repoDetailsPresenter: RepoDetailsPresenter

    var mainInteractor: MainInteractor = mock()
    var repoDetailsView: RepoDetailsView = mock()

    @Before
    fun setup() {
        mockInteractor(mainInteractor)
        repoDetailsPresenter = RepoDetailsPresenterImpl(repoDetailsView, mainInteractor)
    }

    @Test
    fun `requestRepoDetails should return RepoDetails`() {
        //GIVEN
        //NOTHING

        //WHEN
        repoDetailsPresenter.requestRepoDetails(USER_NAME, PROJECT_ID)

        //THEN
        verify(repoDetailsView).showRepoDetails(any<Repo>())
        verify(repoDetailsView, never()).showError(any<String>())
    }

    @Test
    fun `requestRepoDetails when UserName is empty shouldReturnError`() {
        //GIVEN
        //NOTHING

        //WHEN
        repoDetailsPresenter.requestRepoDetails("", PROJECT_ID)

        //THEN
        verify(repoDetailsView, never()).showRepoDetails(any<Repo>())
        verify(repoDetailsView).showError(any<String>())
    }

    companion object {
        private val USER_NAME = "google"
        private val PROJECT_ID = "test"
    }
}
